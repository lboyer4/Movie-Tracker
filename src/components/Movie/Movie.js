import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchUsers } from '../../utils/fetchUsers';
import { toggleFavorite } from '../../actions';
import { Route, Link } from 'react-router-dom';
import MovieDetails from '../MovieDetails/MovieDetails';
import PropTypes from 'prop-types';

export class Movie extends Component {
	constructor(props) {
		super(props);
		this.state = {
			createAccountMsg: '',
			currentMovieId: 0
		}
	}

	handleClick = () => {
		if (!this.props.loggedIn.id)  {
			this.setState({createAccountMsg: '* Please create an account'})
		} else if(this.props.favorited === false) {
			this.props.toggleFavorite(this.props.movie_id)
			this.postFavoriteMovie()
		} else if (this.props.favorited === true) {
			this.props.toggleFavorite(this.props.movie_id)
			this.deleteFavoriteMovie(this.props.movie_id)
		}
	}

	deleteFavoriteMovie = (movie_id) => {
		const { id } = this.props.loggedIn
		
		const url = `http://localhost:3000/api/users/${id}/favorites/${movie_id}`
		const options = {
			method: 'DELETE',
			headers: {
				'Content-Type': 'application/json'
			},
			body: JSON.stringify({
				id,
				movie_id
			})
		}
		fetchUsers(url, options)
	}

	postFavoriteMovie = () => {
		const url = 'http://localhost:3000/api/users/favorites/new';
		const { movie_id, title, poster_path, release_date, vote_average, overview } = this.props
		const user_id = this.props.loggedIn.id
		const options = { 
			method: 'POST',
			headers: {
				'Content-Type': 'application/json'
			},
			body: JSON.stringify({
				movie_id, 
				user_id, 
				title, 
				poster_path, 
				release_date, 
				vote_average, 
				overview 
			})
		}
		fetchUsers(url, options)
	}

	render() {
		let movieId = this.props.movie_id
		let trueMessage = <h4 className='toggle-msg'> Remove Favorite </h4>
		let falseMessage = <h4 className='toggle-msg'> Favorite </h4>
		let toggleMessage = this.props.favorited ? trueMessage : falseMessage
		let message = <h3 className='login-msg'>{this.state.createAccountMsg}</h3>
	  return(
	    <section className='card'>    	
	      <section className='movie-info-wrapper'>
					<Link to=
						{`/movies/${movieId}`}>
						<img src={this.props.poster_path} alt='movie poster'/>
					</Link> 
					<Route exact path=
						{`/movies/${movieId}`} 
						component= { MovieDetails } 
					/>
					<button onClick={this.handleClick} className='fav-btn'> 
						{toggleMessage} 
					</button>
	      </section>
	      {message}
	    </section>
	  )
	}
}

export const mapStateToProps = (state) => ({
	loggedIn: state.loggedIn,
	favorites: state.favorites
})

export const mapDispatchToProps = (dispatch) => ({
	toggleFavorite: (id) => dispatch(toggleFavorite(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(Movie)