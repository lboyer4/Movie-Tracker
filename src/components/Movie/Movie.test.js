import React from 'react';
import { Movie, mapStateToProps,mapDispatchToProps,  } from './Movie';
import { shallow } from 'enzyme';
import { toggleFavorite } from '../../actions';
import { fetchUsers } from '../../utils/fetchUsers';
import { apikey } from '../../apikey';

jest.mock('../../utils/fetchUsers')
fetchUsers.mockImplementation(() =>
	Promise.resolve( {ok: true}) )

describe('Movie', () => {
	let wrapper;

	const mockId = 9;

	const mockMovie = {
		favorited: false,
		movie_id: 514439,
		overview: "When he was 14, Smith drowned in Lake St. Louis and was dead for nearly an hour. According to reports at the time, CPR was performed 27 minutes to no avail. Then the youth's mother, Joyce Smith, entered the room, praying loudly. Suddenly, there was a pulse, and Smith came around.",
		poster_path: "https://image.tmdb.org/t/p/w300_and_h450_bestv2//t58dx7JIgchr9If5uxn3NmHaHoS.jpg",
		release_date: "2019-04-11",
		title: "Breakthrough",
		vote_average: 6.6
	}

	beforeEach(() => {
		wrapper = shallow(
				<Movie 
				{...mockMovie}
					/>
		)
	});

	it('should match snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	describe('handleClick', () => {
		it('should set state with create account messafge if theres no loggedIn id', () => {
			wrapper = shallow(
				<Movie
					loggedIn={{id: null}}
				/>
			)

			wrapper.instance().handleClick()
			expect(wrapper.instance().state.createAccountMsg).toEqual('* Please create an account')
		});

		it('should change favorited to true if favorited is false', () => {
			wrapper = shallow(
				<Movie
					{...mockMovie}
					loggedIn={{id: 1}}
					favorited= {false}
					toggleFavorite={jest.fn()}
				/>
			)

			wrapper.instance().handleClick()
			expect(wrapper.instance().props.toggleFavorite).toHaveBeenCalledWith(mockMovie.movie_id)
		});

		it('should change favorited to false if favorite is true', () => {
			wrapper = shallow(
				<Movie
					{...mockMovie}
					loggedIn={{id: 1}}
					favorited={true}
					toggleFavorite={jest.fn()}
					deleteFavoriteMovie={jest.fn()}
				/>
			)

			wrapper.instance().handleClick()
			expect(wrapper.instance().props.toggleFavorite).toHaveBeenCalledWith(mockMovie.movie_id)
		});
	});

	describe('deleteFavoriteMovie', () => {
		it('should call fetchUsers', () => {
			const wrapper = shallow(
				<Movie 
					{...mockMovie}
					loggedIn={{id: 1}}
					favorited={true}
					toggleFavorite={jest.fn()}
					deleteFavoriteMovie={jest.fn()}
				/>)
			wrapper.instance().deleteFavoriteMovie(mockId)
			expect(fetchUsers).toHaveBeenCalled();
		});
	});

	describe('postFavoriteMovie', () => {
		it('should call fetch users', () => {
			const wrapper = shallow(
				<Movie 
					{...mockMovie}
					loggedIn={{id: 1}}
					favorited={true}
					toggleFavorite={jest.fn()}
				/>
			)

			wrapper.instance().postFavoriteMovie();
			expect(fetchUsers).toHaveBeenCalled();
		});
	});

	describe('mapStateToProps', () => {

		it('should return a user object and a favorites array', () => {

			const mockState = {
				loggedIn: {user: 'name'},
				favorites: ['favorites']
			}

			const expected = {
				loggedIn: mockState.loggedIn,
				favorites: mockState.favorites 
			}

			const mappedProps = mapStateToProps(mockState);

			expect(mappedProps).toEqual(expected);
		});
	});

	describe('mapDispatchToProps', () => {

		it('should call dispatch when using a function from MDTP', () => {

			const mockDispatch = jest.fn();

			const actionToDispatch = toggleFavorite(9);

			const mappedProps= mapDispatchToProps(mockDispatch);

			mappedProps.toggleFavorite(9);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});
	});
});