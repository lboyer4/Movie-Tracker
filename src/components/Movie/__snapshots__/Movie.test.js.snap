// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie should match snapshot 1`] = `
<section
  className="card"
>
  <section
    className="movie-info-wrapper"
  >
    <Link
      to="/movies/514439"
    >
      <img
        alt="movie poster"
        src="https://image.tmdb.org/t/p/w300_and_h450_bestv2//t58dx7JIgchr9If5uxn3NmHaHoS.jpg"
      />
    </Link>
    <Route
      component={[Function]}
      exact={true}
      path="/movies/514439"
    />
    <button
      className="fav-btn"
      onClick={[Function]}
    >
      <h4
        className="toggle-msg"
      >
         Favorite 
      </h4>
    </button>
  </section>
  <h3
    className="login-msg"
  />
</section>
`;
